### PR TITLE
Add patch default and clean up patch

### PR DESCRIPTION
**Pull request by Licenser**
# Pull request

## Description

Sketch of event defaulting as part of #1090. `default` as part of patch is less invasive and problematic then defaulting for the whole expression language since:

1) it is contained to the `patch` operation
2) it has no averse side effects to other possible implementations
3) no new keywords are introduced
4) it mirrors the already existing `merge` operation so is familiar in concept

This PR:

* Add `default => {...}` to patch (mirroring `merge => {...}`)
* Add `default "key" => "value"` (mirroring `merge "key" => {...}` but non recursive)
* Update patch to enforce string keys as part of the gramer instead of creating runtime errors
* Fixes errors for "expected string" to not show a tick instead of a quote
* Moves type errors on patch keys to compile time
* Simplifes the record creation

## Related

* Related Issues: #1090
* Related [docs PR](https://github.com/tremor-rs/tremor-www-docs/pull/168)

## Checklist

* [x] The RFC, if required, has been submitted and approved
* [x] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [x] The code is tested
* [x] Use of unsafe code is reasoned about in a comment
* [x] Update CHANGELOG.md appropriately, recording any changes, bug fixes, or other observable changes in behavior
* [x] The performance impact of the change is measured (see below)

## Performance

A slight performance improvement of ~1% using the `patch` benchmark

### CODE REVIEW

File: tremor-script/src/ast/raw.rs [hunk 1]
```diff
@@ -1286,52 +1260,64 @@ pub enum PatchOperationRaw<'script> {
     /// we're forced to make this pub because of lalrpop
     Insert {
         /// we're forced to make this pub because of lalrpop
-        ident: ImutExprRaw<'script>,
+        ident: StringLitRaw<'script>,
```
Comment (mfelsche) [inline tremor-script/src/ast/raw.rs:1263]: Why do we need to restrict this to literals?
I could imagine to use patch to manipulate a key dynamically by some application logic. For such use cases we'd need an `ImutExprInt` here.

Comment (Licenser) [reply tremor-script/src/ast/raw.rs:1263]: we can still do that using `"#{dynamic_key}"` restricting it to string literals ensurews that we don't have runtime errors for keys not being strings - we can catch this at compiletime. It also prevents confusion when someone writes:

```
patch event of
  insert timestamp => 0
end;
```

which can be easiely be read (and written) as `timestamp` being they key, not `timestamp` being the variable that holds the key.

**Review by mfelsche — DISMISSED** [review]
Looking good.

File: src/codec/json.rs [hunk 2]
```diff
@@ -12,35 +12,60 @@
...
+use std::{cmp::max, marker::PhantomData};
 
 use super::prelude::*;
+use tremor_script::utils::sorted_serialize;
 use tremor_value::AlignedBuf;
 
-pub struct Json {
+/// Sorting for JSON
+pub trait Sorting: Sync + Send + Copy + Clone + 'static {
+    const SORTED: bool;
```
Comment (darach) [inline src/codec/json.rs:23]: Disgusting, I love it, very nice indeed! 🚀

Comment (Licenser) [reply src/codec/json.rs:23]: :D ja and efficient (see https://rust.godbolt.org/z/3nTsd1s1K)

**Review by mfelsche — APPROVED** [review]

**Pull request merged**